### PR TITLE
Feature/tsn 807 settings add an “request deletion

### PR DESCRIPTION
**Pull request by mrilo-sw**

### CODE REVIEW

File: api/models/DownloadBooks.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+	const DownloadBooks = sequelize.define(
```
Comment (frey-sw) [inline api/models/DownloadBooks.js]: Please change name to BooksDownloaded

File: api/resolvers/mutations/Profile/deleteProfile.mutation.js [hunk 2]
```diff
@@ -4,92 +4,76 @@ import { ProfileErrorsEnum } from '../../../enums/errors/ProfileErrorsEnum';
...
+				);
+
+				//delete relations
+				await ProfileRelation.update({ status: ProfileRelationStatusEnum.DEACTIVATED }, { where: { [Op.or]: { ProfileId: input.profileId, relationId: input.profileId } } });
+
+				//delete profile
+				return !!await Profile.update({ status: parseInt(ProfileStatusEnum.DEACTIVATED) }, { where: { id: parseInt(input.profileId) } });
 			} else {
-				throw new ApolloError(ProfileErrorsEnum.NOT_ALLOWED);
+				return !!await Profile.destroy({ where: { id: input.profileId } });
```
Comment (frey-sw) [inline api/resolvers/mutations/Profile/deleteProfile.mutation.js]: Incomplete. Media, comments and everything related to the Profile must be deleted

File: api/resolvers/mutations/User/requestAccountDeletion.mutation.test.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+			);
+		});
+	});
+});
+
+describe('When attributes are correct', () => {
+	beforeEach(async () => {
+		const userModel = models.User;
+		await userModel.destroy({ truncate: { cascade: true } });
+		await models.User.create({
```
Comment (frey-sw) [inline api/resolvers/mutations/User/requestAccountDeletion.mutation.test.js]: Use the generic function generateUser to create fake data

File: api/resolvers/mutations/User/requestAccountDeletion.mutation.test.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+			id: 100,
+			email: 'test2@test.com',
+			password: 'password',
+			role: UserRoleEnum.USER,
+			status: UserStatusEnum.DEACTIVATED,
+			lastLoginDate: new Date(),
+			language: UserLangEnum.en
+		});
+		await models.User.create({
```
Comment (frey-sw) [inline api/resolvers/mutations/User/requestAccountDeletion.mutation.test.js]: Same as previous comment

File: api/resolvers/mutations/User/requestAccountDeletion.mutation.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import { UserErrorsEnum } from '../../../enums/errors/UserErrorsEnum';
+import { UserStatusEnum } from '../../../enums/User';
+
+const requestAccountDeletion = {
+	requestAccountDeletion: async (root, { input }, { models: { User } }) => {
+		try {
+
+			// Find user
+			const user = await User.findOne({ where: { id: input.userId } });
+			if (!user) return true;
```
Comment (frey-sw) [inline api/resolvers/mutations/User/requestAccountDeletion.mutation.js]: Fix return error

File: api/resolvers/mutations/User/requestAccountDeletion.mutation.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+			if (user.status === UserStatusEnum.DEACTIVATED) throw new ApolloError(UserErrorsEnum.ACCOUNT_DEACTIVATED);
```
Comment (frey-sw) [inline api/resolvers/mutations/User/requestAccountDeletion.mutation.js]: Also check for user deleted status

File: api/resolvers/queries/SatisfactionSurvey/viewSurvey.query.test.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+			{
+				id: 112,
+				email: '6@yopmail.com',
+				password: 'password',
+				role: UserRoleEnum.USER,
+				status: UserStatusEnum.ACTIVE,
+				lastLoginDate: new Date(),
+				language: UserLangEnum.en
+			}
+			]);
```
Comment (frey-sw) [inline api/resolvers/queries/SatisfactionSurvey/viewSurvey.query.test.js]: Use generic user creation functions

File: web/src/components/Layout/Admin/SurveyResults/SurveyResults.js [hunk 8]
```diff
@@ -127,18 +133,31 @@ const SurveyResults = () => {
...
+			{ key: t('SURVEY.FAVORITES.FAMILY'), value: (rawFavoriteFeatures.length ? responsePercentage(familyTreeResponses.length, rawFavoriteFeatures.length) : 0) + '%' },
+			{ key: t('SURVEY.FAVORITES.WRITING'), value: (rawFavoriteFeatures.length ? responsePercentage(thoughtsResponses.length, rawFavoriteFeatures.length) : 0) + '%' },
+			{ key: t('SURVEY.FAVORITES.OTHER'), value: (rawFavoriteFeatures.length ? responsePercentage(otherResponses.length, rawFavoriteFeatures.length) : 0) + '%' }
 		];
 		return (
 			<>
 				<div className='divider marginTop_1 marginBottom_2'></div>
 				<NumberedTitle number={3} title={t('SURVEY.FAVORITES.QUESTION')} headingSmall>
-					<DataResults items={favoriteFeaturePercentageArray} dataStyle={DATARESULTS_STYLE.VERTICAL} />
+					<DataResults insideButton items={favoriteFeaturePercentageArray.sort(compareValues('value', 'desc'))} dataStyle={DATARESULTS_STYLE.VERTICAL}>
```
Comment (frey-sw) [inline web/src/components/Layout/Admin/SurveyResults/SurveyResults.js]: Too much logic on items attribute, please assign a variable to it

File: web/src/components/Layout/Admin/SurveyResultsAnswers/SurveyResultsAnswers.js [hunk 9]
```diff
@@ -12,45 +14,164 @@ import FlexContainer, { FLEX_DIR, FLEX_ALIGN, FLEX_JUSTIFY } from 'UI/FlexContai
...
+					limit: paginationState.pageSize,
+					answerType: columnToFilter,
+					searchText: searchTerm
+				}
+			},
+			fetchPolicy: 'network-only'
+		});
+	};
+
+	if (!called && paginationState.search === undefined) getAnswers();
```
Comment (frey-sw) [inline web/src/components/Layout/Admin/SurveyResultsAnswers/SurveyResultsAnswers.js]: paginationState.search === undefined simplified to !paginationState.search

File: web/src/components/Layout/User/RequestDeleteModal/RequestDeleteModal.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import { useMutation } from '@apollo/react-hooks';
+import PropTypes from 'prop-types';
+import { useTranslation } from 'react-i18next';
+
+import REQUEST_DELETION_MUTATION from 'core/mutations/User/REQUEST_DELETION_MUTATION';
+import Button, { BUTTON_COLOR, BUTTON_STYLE, BUTTON_ICON_POS } from 'UI/Button/Button';
+import FlexContainer, { FLEX_JUSTIFY, FLEX_ALIGN, FLEX_DIR } from 'UI/FlexContainer/FlexContainer';
+import Modal from 'UI/Modal/Modal';
+import { faTimes, faCheck, faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
```
Comment (frey-sw) [inline web/src/components/Layout/User/RequestDeleteModal/RequestDeleteModal.js]: Move imports to the top

File: web/src/components/Layout/User/Settings/Settings.js [hunk 11]
```diff
@@ -60,19 +65,29 @@ const Settings = () => {
...
+									actionClick={deleteAccountModalHandler}
+									actionText={t('SETTINGS.ACCOUNT.REQUEST_DELETE')}
+									actionIcon={faUserTimes}
+								/>
 							</SettingsSection>
 						</Fade>
 						{
 							called && !loading && data ?
-								<ProfileDeactivation user={data.user} />
+								<ProfileDeactivation setSettingsState={() => setSettingsState({ ...settingsState, setProfile: false })} settingsState={settingsState} user={data.user} />
```
Comment (frey-sw) [inline web/src/components/Layout/User/Settings/Settings.js]: Move setSettingsState to a function

File: web/src/components/Layout/User/Settings/Settings.js [hunk 12]
```diff
@@ -60,19 +65,29 @@ const Settings = () => {
...
 								: null
 						}
 					</div>
 				</FlexContainer>
 
 			</div>
 
-			{data && changeEmail ? <ChangeEmailModal user={data.user} close={emailModalHandler} /> : null}
+			{data && settingsState.changeEmailModal ? <ChangeEmailModal user={data.user} close={emailModalHandler} /> : null}
+			{data && settingsState.requestDeleteModal ? <RequestDeleteModal setProfile={() => setSettingsState({ ...settingsState, setProfile: true, requestDeleteModal: false })} user={data.user} close={deleteAccountModalHandler} /> : null}
```
Comment (frey-sw) [inline web/src/components/Layout/User/Settings/Settings.js]: Move setSettingsState to a function

File: web/src/core/mutations/User/REQUEST_DELETION_MUTATION.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+import { gql } from 'apollo-boost';
+
+export default gql`
+	mutation RequestAccountDeletion($input: RequestDeletionInput!) {
+		requestAccountDeletion(input: $input)
```
Comment (frey-sw) [inline web/src/core/mutations/User/REQUEST_DELETION_MUTATION.js]: Response must be added to control errors in the frontend

File: api/resolvers/mutations/Profile/deleteProfile.mutation.js (hunk 2 above)
Comment (mrilo-sw) [reply api/resolvers/mutations/Profile/deleteProfile.mutation.js]: Added big comment explaining that this mutation will be user for 2 different purposes, and the condition for each one

Comment (mrilo-sw) [reply api/resolvers/mutations/Profile/deleteProfile.mutation.js]: > Incomplete. Media, comments and everything related to the Profile must be deleted

Added big comment explaining that this mutation will be user for 2 different purposes, and the condition for each one

File: api/models/DownloadBooks.js (hunk 1 above)
Comment (tmassuia-sw) [reply api/models/DownloadBooks.js]: Still the same name like we chat

**Pull request merged**